src/backend/parser/parse_agg.c

```diff
@@ -1505,16 +1536,153 @@ get_groupclause_exprs(Node *grpcl, List *targetList)
 		List *exprs = (List *)grpcl;
 		ListCell *lc;
 
-		foreach (lc, exprs)
+		foreach(lc, exprs)
 		{
-			result = list_concat(result, get_groupclause_exprs((Node *)lfirst(lc),
+			result = list_concat(result, get_groupclause_tles((Node *)lfirst(lc),
 															   targetList));
 		}
 	}
 
 	return result;
 }
 
+/*
+ * get_com_grouping_ressortgroupref_routine -
+ *     Return a list of common ressortgrouprefs of the current grouping
+ *     expression recursively. It is assumed that attributes used in all
+ *     groupings are already present in the current grouping. Otherwise,
+ *     missing references are removed from the com_refs list.
+ */
+static List*
+get_com_grouping_ressortgroupref_routine(Node *grpcl, List *targetList, List *com_refs)
+{
+	List *result = NIL;
+
+	if ( !grpcl )
+		return result;
+
+	Assert(IsA(grpcl, SortGroupClause) ||
+		   IsA(grpcl, GroupingClause) ||
+		   IsA(grpcl, List));
+
+	if (IsA(grpcl, SortGroupClause))
+	{
+		TargetEntry *tle = get_sortgroupclause_tle((SortGroupClause *) grpcl, targetList);
+		result = lappend_int(result, tle->ressortgroupref);
+	}
+	else if (IsA(grpcl, GroupingClause))
+	{
+		ListCell *l;
+		GroupingClause *gc = (GroupingClause*)grpcl;
+
+		/* Cube and Rollup do not contain common attributes */
+		if (gc->groupType == GROUPINGTYPE_CUBE || gc->groupType == GROUPINGTYPE_ROLLUP)
+		{
+			list_free(com_refs);
+			return NIL;
+		}
+
+		foreach(l, gc->groupsets)
+		{
+			List *grouping_refs = get_com_grouping_ressortgroupref_routine((Node*)lfirst(l), targetList, com_refs);
+			
+			/* An empty grouping has no common attributes */
+			if (!grouping_refs)
+				return NIL;
+
+			/*Exclude refs that did not appear in this grouping*/
+			ListCell *lc = list_head(com_refs);
+			while (lc)
+			{
+				if (list_member_int(grouping_refs, lfirst_int(lc)))
+				{
+					lc = lnext(lc);
+				}
+				else
+				{
+					ListCell *lc_del = lc;
+					lc = lnext(lc);
+					com_refs = list_delete_int(com_refs, lfirst_int(lc_del));
+				}
+			}
```

Comment:
I suggest to backport `list_intersection_int()` from 'adb-7.2.0' and simpify this code 
```suggestion
			com_refs = list_intersection_int(com_refs, grouping_refs);
```

Reply:
Done